### ISSUE TITLE
Auth: Another TSIG noncompliance with RFC2845

### ISSUE THREAD

**Issue by Enna1** [PowerDNS/pdns#9104]
- Program: Authoritative
 - Issue type: Bug report

### Short description

When pdns auth server receives a query with TSIG RR, and server time is outside the time interval specified by the request (which is: TSIG Time Signed, plus/minus TSIG Fudge), the server generates a response with RCODE 9 (NOTAUTH) but no TSIG RR.
This is a noncompliance with RFC2845, as RFC2845 says:

> 4.5.2. TIME check and error handling
If the server time is outside the time interval specified by the request (which is: Time Signed, plus/minus Fudge), the server MUST generate an error response with RCODE 9 (NOTAUTH) and TSIG ERROR 18 (BADTIME).
...
A response indicating a BADTIME error MUST be signed by the same key as the request. It MUST include the client’s current time in the time signed field, the server’s current time (a u_int48_t) in the other data field, and 6 in the other data length field.

### Environment

 - Operating system: Ubuntu18.04
 - Software version: PowerDNS Authoritative Server 4.3.0
 - Software source: Compiled myself 

### Expected behaviour

A TSIG signed response with RCODE 9 (NOTAUTH) and TSIG ERROR 18 (BADTIME). And TSIG other data length field is 6, TSIG other data field is server’s current time.
The follwing pic is bind's response for the same query
![bind-tsig-outside-timeinterval](https://user-images.githubusercontent.com/33259172/81894806-4e2e6680-95e3-11ea-87c4-0db5dd88c101.png)

### Actual behaviour

A response with RCODE 9 (NOTAUTH) but no TSIG RR.
The follwing pic is PowerDNS Authoritative Server 4.3.0's response
![pdns-tsig-outside-timeinterval](https://user-images.githubusercontent.com/33259172/81894886-7027e900-95e3-11ea-8e93-c3c07e017b3c.png)

**Comment by rgacogne** [comment]
Out of curiosity, is the current behavior causing a real interoperability issue? I'm usually all for sticking to the RFC but generated a signed response to an invalid query is a bit error-prone. Not doing so has saved us from a few CVEs already, with CVE-2017-3143 coming immediately to mind. So in my humble opinion if we fix this issue we should be careful to stick to https://tools.ietf.org/html/draft-ietf-dnsop-rfc2845bis-08 instead of following the original rfc2845.

**Comment by Enna1** [comment]
Hi @rgacogne 

I don't know if current behavior is causing a real interoperability issue.

A few days ago, rfc2845bis (https://datatracker.ietf.org/doc/draft-ietf-dnsop-rfc2845bis) has been released as a Internet Standard RFC8945 (https://tools.ietf.org/html/rfc8945).
RFC8945 obsoletes 2845, so I think we should stick to RFC8945.

As RFC8945 says:

> TSIG was originally specified by [RFC2845].  In 2017, two name server implementations strictly following that document (and the related [RFC4635]) were discovered to have security problems related to this feature ([CVE-2017-3142], [CVE-2017-3143], [CVE-2017-11104]).  The implementations were fixed, but to avoid similar problems in the future, the two documents were updated and merged, producing this revised specification for TSIG.

> To validate the received TSIG RR, the server MUST perform the following checks in the following order:
> 1.  Check key
> 2.  Check MAC
> 3.  Check time values
> 4.  Check truncation policy

In RFC2845, the checking order is: check KEY, check TIME values, check MAC.

The checking order described in RFC8945 means, when we check time values, we has checked MAC and TSIG MAC is verified successfully. Then if the server time  is outside the time interval specified by the request (which is: Time Signed, plus/minus Fudge), we generate a TSIG signed response with RCODE 9 (NOTAUTH) and TSIG ERROR 18 (BADTIME). And TSIG other data length field is 6, TSIG other data field is server’s current time.

So I think doing the TSIG validation as RFC8945 described will not cause any security problem.

**Comment by rgacogne** [comment]
Thanks for the update! I'm afraid I'm still not convinced there is any gain to changing the existing code unless there is an issue with our current behaviour, but other might feel differently.

**Comment by Habbie** [comment]
I agree with @rgacogne - I prefer a minor incompliance if it reduces attack surface, so I'm closing this.
